src/libraries/Microsoft.PowerFx.Connectors/Public/CdpTableValue.cs

```diff
@@ -72,33 +75,78 @@ public void Refresh()
+            if (expectedRT is not AggregateType) 
+            {
+                if (value.Type is not AggregateType)
+                {
+                    var expectedValue = ConvertToExpectedType(expectedRT, value);
+                    return expectedValue;
+                }
+                else if (value.Type is RecordType resultRT)
+                {
+                    var expectedValue = await ExtractResultAsync((RecordValue)value, parameters, cancel).ConfigureAwait(false);
+                    return expectedValue;
+                }
+                else
+                {
+                    throw new InvalidOperationException($"Invalid result type {value.Type}, expected type was {expectedRT}");
+                }
+            }
+            else
+            {
+                throw new InvalidOperationException($"Invalid expected type {expectedRT} for {nameof(ExecuteQueryAsync)}");
+            }
         }
-    }   
 
-    internal static class ODataParametersExtensions
-    {
-        public static ODataParameters ToOdataParameters(this DelegationParameters parameters)
+        private static async Task<FormulaValue> ExtractResultAsync(RecordValue value, DelegationParameters parameters, CancellationToken cancellationToken)
         {
-            DelegationParameterFeatures allowedFeatures = 
-                DelegationParameterFeatures.Filter | 
-                DelegationParameterFeatures.Top | 
-                DelegationParameterFeatures.Columns | // $select
-                DelegationParameterFeatures.Sort;     // $orderby
+            var valueTable = (TableValue)(await value.GetFieldAsync(DelegationParameters.ODataResultFieldName, cancellationToken).ConfigureAwait(false));
+            if (valueTable.Rows.Count() != 1)
+            {
+                throw new InvalidOperationException("value Table should always have 1 rows for aggregation result");
+            }
 
-            parameters.EnsureOnlyFeatures(allowedFeatures);
+            var valueRecord = valueTable.Rows.First().Value;
+            var result = await valueRecord.GetFieldAsync(DelegationParameters.ODataAggregationResultFieldName, cancellationToken).ConfigureAwait(false);
+            result = ConvertToExpectedType(parameters.ExpectedReturnType, result);
+            return result;
+        }
+
+        private static FormulaValue ConvertToExpectedType(FormulaType expectedType, FormulaValue value)
+        {
+            var valueType = value.Type;
 
-            ODataParameters op = new ODataParameters()
+            if (value is BlankValue)
             {
-                Filter = parameters.GetOdataFilter(),
-                Top = parameters.Top.GetValueOrDefault(),
-                Select = parameters.GetColumns(),
-                OrderBy = parameters.GetOrderBy()
-            };
+                return value;
+            }
 
-            return op;
+            if (expectedType == valueType)
```

Comment:
>    if (expectedType == valueType) [](http://example.com/codeflow?start=8&length=34)

Depending on how much coercion is needed, we do have some helpers that can help here.

Reply:
[https://github.com/microsoft/Power-Fx/blob/main/src/libraries/Microsoft.PowerFx.Core/Public/Values/PrimitiveValueConversions.cs](https://github.com/microsoft/Power-Fx/blob/main/src/libraries/Microsoft.PowerFx.Core/Public/Values/PrimitiveValueConversions.cs)